test/unit/mongo_logger.test.ts

```diff
@@ -1276,6 +1286,218 @@ describe('class MongoLogger', function () {
+      context('when stream is not stderr', function () {
+        let stderrStub;
+
+        beforeEach(function () {
+          stderrStub = sinon.stub(process.stderr);
+        });
+
+        afterEach(function () {
+          sinon.restore();
+        });
+
+        context('when stream user defined stream and stream.write throws async', async function () {
+          it('should catch error, not crash application, warn user, and start writing to stderr', async function () {
+            const stream = {
+              async write(_log) {
+                await new Promise(r => setTimeout(r, 500));
+                throw Error('This writable always throws, but only after at least 500ms');
+              }
+            };
+            const logger = new MongoLogger({
+              componentSeverities,
+              maxDocumentLength: 1000,
+              logDestination: stream
+            });
+            // print random message at the debug level
+            logger.debug('random message');
+
+            // before timeout resolves, no error
+            expect(stderrStub.write.getCall(0)).to.be.null;
+
+            // manually wait for timeout to end
+            await new Promise(r => setTimeout(r, 600));
+
+            // stderr now contains the error message
+            let stderrStubCall = stderrStub.write.getCall(0).args[0];
+            stderrStubCall = stderrStubCall.slice(stderrStubCall.search('c:'));
+            expect(stderrStubCall).to.equal(
+              `c: 'client', s: 'error', message: 'User input for mongodbLogPath is now invalid. Now logging to stderr.', error: 'This writable always throws, but only after at least 500ms' }`
+            );
+
+            logger.debug('random message 2');
+            const stderrStubCall2 = stderrStub.write.getCall(1);
+            expect(stderrStubCall2).to.not.be.null;
+          });
+        });
+      });
+    });
+    context('when async stream has multiple logs with different timeouts', async function () {
+      it('should preserve their order', async function () {
+        const stream = {
+          buffer: [],
+          async write(log) {
+            if (log.c === 'longer timeout') {
+              await new Promise(r => setTimeout(r, 2000));
+            } else if (log.c === 'shorter timeout') {
+              await new Promise(r => setTimeout(r, 500));
+            }
+            this.buffer.push(log.c);
+          }
+        };
+        const logger = new MongoLogger({
+          componentSeverities,
+          maxDocumentLength: 1000,
+          logDestination: stream
+        });
+
+        logger.debug('longer timeout');
+        logger.debug('shorter timeout');
+        logger.debug('no timeout');
+
+        expect(stream.buffer.length).to.equal(0);
+
+        await new Promise(r => setTimeout(r, 2100));
+        expect(stream.buffer.length).to.equal(1);
+        expect(stream.buffer[0]).to.equal('longer timeout');
```

Comment:
Love this test!

Small improvement, instead of checking length and the item at an offset can you deep equal check with a hardcoded array?

```js
expect(stream.buffer).to.deep.equal(['longer timeout']);
```
And same below but with the subsequent logs. Then the test will print out the whole array if something goes wrong.